text/1004-transactions-api.rst

```diff
@@ -651,16 +945,129 @@ There are few problems of this approach:
    network failures would be an issue
 
 
-Keep `transaction` as but Add a Helper for Retrying
----------------------------------------------------
+Keep `transaction` as is but Add a Helper for Retrying
+------------------------------------------------------
 
-The problem with this approach is that it hard to teach using `retry`
+The problem with this approach is that it hard to teach using ``retry``
 when raw transactions "work on my laptop". However, this is somewhat
 alleviated by failure injection.
 
 
+Retry All Single Queries
+------------------------
+
+This specification describes that read-only non-transactional queries
+should be retried automatically.
+
+While it's tempting to retry all ``pool.query`` and ``pool.execute``
+calls (even modifications), it **gives the false sense of security**: no
+"concurrent update" issues are seen.  But it's better to see such error
+and turn the whole block of code into a transaction rather than just a
+mutation. I.e. retrying a single mutable request on a "concurrent
+update" error must be a deliberate decision.
+
+
+Is ``with_modifications`` Needed?
+---------------------------------
+
+It's intuitive that databases are mutable by default. But there is a
+large class of applications that are mostly read-only and must have
+limited and easy to find places having mutations. For those apps it's
+better to have read-only connection by default and use a pattern like
+this for writes:
```

Comment:
The RFC starts with stating and showing examples of `edgedb.connect()` returning a mutable connection. So I assume that this section advocates for adding a way to convert a read-only connection (obtained via `edgedb.connect().read_only()`) to a mutating one?  I'm confused.

Reply:
Yeah. Paul, we've argued over this a lot already, and I see no end to this. So I propose we do this instead: we don't implement `with_modifications()` right now. We can always add this later, it's really not a critical feature and we can add it at any time when we get some user feedback.

I'd really like this RFC to only implement essential and **non-controversial** APIs, where non-controversial means unanimous support of you, Elvis, and I.

Reply:
> So I assume that this section advocates for adding a way to convert a read-only connection (obtained via edgedb.connect().read_only()) to a mutating one?

Yes, in the top of the application, you either do (assuming `aiohttp`):
```python
app['edgedb'] = await create_async_pool('name')
```
Or you do:
```python
app['edgedb'] = (await create_async_pool('name')).read_only(allow_upgrade=True)
```
In the latter case, you prevent arbitrary functions execute side-effects unless run within `with_modification().retry()` transaction and transaction is passed to the function directly.

> In my mind, the main feature and premise of read_only() is that it allows better scalability by allowing connections to standby replicas.

Well:
1. We don't implement that scalability just yet anyway
2. With regards of this RFC the original premise of read-only was retrying queries automatically. But well, it's solved by `HAS_FEATURES` so is not strictly needed
3. I don't see anything in `with_modifications` as controversial (except the name). Controversial would be official recommendation of this, but not the support. And in previous discussion I think it was agreed as "yes we can support this pattern".

Anyway I'm dropping the function and move this section to the Future Work, so that we can finally move forward.